Encrypt PBEM/PBF passwords when stored (#1884)

ssoloff opened the pull request:
This PR should fully address #1884.  It builds on #1929 such that, if the user chooses to have TripleA remember their PBEM/PBF password, the password will be encrypted.

#### Functional changes
* PBEM/PBF passwords are encrypted if they are saved to the local cache.  See the Security Review below for details of the implementation.
* When selecting the "Remember my password" checkbox, the message warning the user their password will be stored unencrypted is no longer displayed.

#### Functional issues for review
None.

#### Refactoring changes
None.

#### Testing
I went through pretty much the same set of tests I performed in #1929.  However, it would be appreciated if someone else can verify this feature works as expected (especially outside the US, just in case there's some unexpected crypto behavior).

#### Security review

##### Overview

A new class, `PasswordManager`, was added to manage a user's PBEM/PBF passwords.  Both `GenericEmailSender` and `AbstractForumPoster` delegate to `PasswordManager` to protect and unprotect a password stored in `local.cache`.

##### Master password

`PasswordManager` uses a user-specific randomly-generated master password (MP) to protect the PBEM/PBF passwords.  The MP will be automatically created if it does not already exist.  The MP is stored unencrypted in the user's preference tree (HKCU on Windows; `~/.java/.userPrefs` on Linux; unknown on OSX).  This preference tree is locked down such that only the user can read/write its contents.  Therefore, the MP is protected from unintentional disclosure, but it is still readable by malware running as the user.

The MP is generated using 21 bytes from a secure SHA1PRNG, which is then encoded into a 28-character Base64 string.  Per [this table](https://en.wikipedia.org/wiki/Password_strength#Random_passwords), 27 Base64 characters (case insensitive alphanumeric) should be enough to generate 160 bits of entropy.

The MP is stored as a `byte[]` in the user preference tree under the node `games/strategy/engine/pbem` in the key `PASSWORD_MANAGER_MASTER_PASSWORD`.

##### PBEM/PBF password protection

The `PasswordManager#protect()` method is used to protect a PBEM/PBF password using the following algorithm:

1. Decode the password into a UTF-8-encoded array of bytes (this is the plaintext).
1. Create 64 bytes of salt from a secure SHA1PRNG.  64 bytes was chosen per [this reference](https://crackstation.net/hashing-security.htm#salt) because SHA-512 is used as the hashing function in the key-generation function (see below).
1. Create an AES key with the PBKDF2WithHmacSHA512 key generator using the MP, salt, an iteration count of 20,000, and a key length of 128 bits.
    * I based the iteration count on information from [this post](https://security.stackexchange.com/questions/3959/recommended-of-iterations-when-using-pkbdf2-sha256) and by measuring key generation times using different iteration counts on my machine (which isn't necessarily representative of the user base).  20,000 seemed to be fast enough and agrees with the recommended range in the above post and linked posts.
    * Oracle JRE 8 is limited to 128-bit AES keys without the JCE Unlimited Strength Jurisdiction Policy installed on the end-user's machine.  I'm assuming this is an unacceptable requirement for TripleA users.  Oracle JRE 9 and OpenJDK (all versions) do not have this restriction.  So we can upgrade to AES-256 at a later time, if required.
1. Encrypt the plaintext using the key generated from the previous step (this is the ciphertext).
1. Base64 encode the ciphertext and the salt.  The "protected password" is the string `<base64-salt>.<base64-ciphertext>`.

The `PasswordManager#unprotect()` method is used to unprotect a previously-protected PBEM/PBF password using the following algorithm:

1. Extract the Base64-encoded ciphertext and salt from protected password and Base64-decode them.
1. Create an AES key using the MP and salt with the same algorithm that was used above for encryption.
1. Decrypt the ciphertext using the key generated from the previous step (this is the plaintext).
1. Encode the UTF-8-encoded plaintext into a string.

##### Serialization and deserialization

As stated above, `GenericEmailSender` and `AbstractForumPoster` delegate to `PasswordManager` to generate the protected password before serialization and to generated the unprotected password after deserialization.  They do this by using the `readObject()` and `writeObject()` hooks in the serialization framework.

Backwards compatibility is handled by assuming passwords are stored unprotected.  Once the new code runs, the serialized format holds a flag indicating the passwords are now protected, and thus, they should be unprotected during the next deserialization.

If an error occurs while protecting or unprotecting a password, the corresponding `m_password` field in these classes is set to the empty string and the error is logged.

* The exception is not re-thrown during deserialization because there's really nothing that can be done at that point, and the user will just have to re-enter their password.
* The exception is not re-thrown during serialization because this triggers an undesirable side effect: the serialization of the entire local bean cache is aborted and the file is deleted.  Thus, all the user's passwords would be lost in the event an error occurred while protecting just a single password.  Again, it seems like a good compromise to just have the user re-enter their password in this scenario.

ssoloff commented on src/main/java/games/strategy/engine/pbem/PasswordManager.java:
```diff
@@ -0,0 +1,366 @@
+  }
+
+  @VisibleForTesting
+  static char[] newMasterPassword() throws GeneralSecurityException {
```
This is the core of master password generation.

ssoloff commented on src/main/java/games/strategy/engine/pbem/PasswordManager.java:
```diff
@@ -0,0 +1,366 @@
+   *
+   * @see #unprotect(String)
+   */
+  String protect(final char[] unprotectedPassword) throws PasswordManagerException {
```
This is the core of PBEM/PBF password protection.

ssoloff commented on src/main/java/games/strategy/engine/pbem/PasswordManager.java:
```diff
@@ -0,0 +1,366 @@
+   *
+   * @see #protect(char[])
+   */
+  char[] unprotect(final String protectedPassword) throws PasswordManagerException {
```
This is the core of PBEM/PBF password unprotection.

RoiEXLab commented:
I briefly skimmed through the code, looks good for the most part.
I will do a complete review later thie day

panther2 commented:
@ssoloff 
Any chance to encrypt the stored user name, too?

Use case: For example on axisandallies.org the user name (login-name) may be different from the (displayed) forum name.

ssoloff commented:
@panther2 I actually considered including the user name as part of the "secret data" to be optionally stored when I submitted #1929, but I didn't think it was important at the time.  Now that you've asked, though... :smile:

It should be relatively simple to encrypt the username, as well.  What would require more work is if the user name needs to be optionally stored.  That is, convert the "remember my password" option to "remember me."  That's probably outside the scope of this PR, but if you feel such a feature is important, please open a new issue to be addressed once this PR and #1884 are closed.

Let me prototype encrypting the user name later today, and I'll get back to you if it looks to be problematic.  The biggest change will probably be renaming the `PasswordManager` class to something like `SecretManager`.

ssoloff commented:
@panther2 User names are now encrypted.  Thanks for your input.  I will probably open a separate PR after this one is merged to allow the user to opt-out of having TripleA remember their user name in addition to their password.

For reviewers, the following have been changed:

* `GenericEmailSender` and `AbstractForumPoster` now protect the user name, as well as the password.
* `PasswordManager*` has been renamed to `CredentialManager*`.
* The preference key where the master password is stored has been renamed to `CREDENTIAL_MANAGER_MASTER_PASSWORD`.

panther2 commented:
Great, thanks a lot!

ssoloff commented:
@panther2 Since you brought up the use case about not wanting to disclose your user name (e.g. forum login), you probably want to know this...  That user name is stored in the save game file you email/post to your opponent(s).  The PBEM case isn't a big deal because the recipient already knows your email by the fact that you sent it.  However, if I understand your concern, there can be a loss of privacy in PBF because other forum users may not have knowledge of your forum user name, but they can now discover it by examining the save game.

When I add the option to not store the user name locally, I'll also ensure that it is always scrubbed from the save game just as the password is.

panther2 commented:
@ssoloff Thanks for the update. I never realized that the forum user name is stored in the tsvg-savegame. 
I never saw it there and cannot find it there, I have to admit...

When I open a PBF-savegame the only information taken over to the TripleA-PBF-setup-screen is the email data.

The forum data is taken from my own local file. I cannot even see my oponents forum data anywhere. So why are the forum user names stored in the savegame? This appears to be pointless.

Just wondering...

ssoloff commented:
@panther2 First, you need to `gunzip` the .tsvg file because it is compressed.  Then you can just `grep` for your email address.  For example:

```
$ gunzip --stdout triplea.tsvg > triplea.tsvg.ser
$ grep "my-forum-name" triplea.tsvg.ser
Binary file triplea.tsvg.ser matches
```

I just tested this on forums.triplea-game.org and verified that my user name is present in the posted save game file.

panther2 commented:
@ssoloff Indeed, I was able to reproduce it now, thanks. But again: Why do we need the forum user name inside the savegame - assuming that the forum account information is provided by the local file?

All my opponent needs from the savegame is the mail-address. Why would he need my forum-account name? He needs his own name, which he gets from his local file...

ssoloff commented on src/main/java/games/strategy/engine/pbem/CredentialManager.java:
```diff
@@ -0,0 +1,366 @@
+ */
+final class CredentialManager implements AutoCloseable {
+  private static final String CIPHER_ALGORITHM = "AES";
+  private static final Charset MASTER_PASSWORD_CHARSET = StandardCharsets.ISO_8859_1;
```
Good question.  The master password only consists of Base64 characters (`[a-zA-Z0-9+/=]`).  Obviously, in this case, it would be equivalent to use either UTF-8 or ISO-8859-1, but I chose the latter because that's what the Oracle `Base64` encoder/decoder classes use (see, for example, the Javadocs for [`Base64$Encoder#encodeToString()`](https://docs.oracle.com/javase/8/docs/api/java/util/Base64.Encoder.html#encodeToString-byte:A-).

ssoloff replied:
Actually, thanks for making me think about this again, @RoiEXLab.  In the future, we may choose to allow the user to specify the master password, as the `CredentialManager` code is already set up to handle this.  In that case, they should be allowed to use any Unicode character.  Just because we chose to internally generate a master password consisting of only Base64 characters shouldn't preclude the user's choice of characters in the future.  So I will just go ahead and change it to UTF-8 to avoid having to think about this again. :smile:

ssoloff commented:
@panther2 Unfortunately, I don't know the original reason for persisting this information to the save game file.  However, I can theorize that is was just to simplify the code.  Basically, the same object (which contains all of the information you enter in the UI) is saved to both the local cache and the save game (with the exception that the code removes the password before storing the object in a save game file).  Ideally, there would be two different representations for each medium with the minimum amount of information required by each.

ssoloff commented:
@RoiEXLab I updated `CredentialManager` to use UTF-8 encoding for the master password.  I first verified the unit tests failed when using a master password containing Unicode characters, and then passed after I switched the charset from ISO-8859-1 to UTF-8.

ssoloff commented:
@RoiEXLab That's basically what the "salt" is being used for here.  If I didn't salt the key generation, then encrypting the same plaintext using the same master password would produce identical ciphertext, thus making the user's `local.cache` file susceptible to a dictionary attack.

I agree it may be conceptually different from how you traditionally see a salt being applied with a hash function.  But, in reality, that's what's actually happening behind the scenes.  PBKDF2 is a key-generation function that uses a hash function (SHA-512 in our case) applied to the master password and the salt to produce the symmetric key that is used to encrypt and decrypt a single credential.  So, if you take a step back to that point, it should be clear that the salt is being used with a hash function just as you expect. :smile:

ron-murhammer commented:
@ssoloff So will this be able to handle the existing username/passwords? Or will all users have to re-enter username/password (this would include existing or new games since this is stored locally)?

ssoloff commented:
@ron-murhammer Yes, it should all Just Work.  The code handles reading the current unencrypted format for `GenericEmailSender` and `AbstractForumPoster` but will always write back the encrypted values.  Thus, you can't go back to an old engine version and expect the encrypted values to be available (you'll get a bunch of base64-encoded gobbledygook if you do).

This is the first test case I ran through to ensure an old `local.cache` is handled properly:

* Start TripleA 1.9.0.0.3635.
* Select PBEM/PBF game and enter configuration for AxisAndAllies.org and Gmail.
* Play game.
* Exit TripleA.
* Verify unencrypted credentials are present in `local.cache`.
* Start TripleA from this branch.
* Select PBEM/PBF game and verify configuration for AxisAndAllies.org and Gmail are present and match previous values.
* (Optional) Select "Remember my password" in either or both PBF and PBEM configurations.
* Play game.
* Exit TripleA.
* Verify encrypted credentials are present in `local.cache`.
    * Note that if you did not select "Remember my password", an encrypted password will still be present in `local.cache`, but it's unencrypted value is just the value of `USE_TRANSITIVE_PASSWORD`.

Then I went through a PBEM game test case to ensure an old save game is handled properly (the behavior should be identical for PBF since the `local.cache` test worked):

* Start TripleA 1.9.0.0.3635.
* Select PBEM/PBF game and enter configuration for Gmail.
* Play game.
* Send save game via email.
* Exit TripleA.
* Start TripleA from this branch.
* Open saved game that was received via email and verify no errors in console.
* Select PBEM/PBF game and ensure configuration is correct.
* Play game and verify no errors in console.
* Send second save game via email.
* Exit TripleA.
* Start TripleA from this branch.
* Open second saved game that was received via email and verify no errors in console.
* Select PBEM/PBF game and ensure configuration is correct.
* Play game and verify no errors in console.
* Exit TripleA.

Are there any other scenarios I need to be concerned about?

ssoloff commented:
@ron-murhammer I may have to run an additional test because a possibility is just dawning on me...  My second test case above (the save game) may only work because I opened all the save games on the same machine (i.e. with the same master password).

I believe the code is still going to attempt to decrypt the encrypted credentials in the save game, but it will fail because, in a real situation, the player who received the save game will have a different master password.  However, it **should not** affect the user because we pretty much ignore the failure and keep going (we do quietly log it, though).

Let me test this scenario to make sure there's no hidden issue there...

ssoloff commented:
@ron-murhammer Yeah...  So that's a problem.  As expected, the decrypt fails and we log it silently.  But the real problem is that the user's credentials are cleared out, so that, yes, they would have to re-enter them.

I guess I'm a bit confused.  I'm not sure why the local user's PBEM/PBF credentials would come from the save game because they will always be the remote user's (albeit with a dummy password).  I just assumed the credentials were  coming from `local.cache` and the only thing the save game was providing was information like the forum topic, the email subject, etc.

Let me test the same scenario between two 3635 clients.  I'll need to think about this a bit.  It may require a more complex solution than I had originally thought.

ron-murhammer commented:
@ssoloff Ok. Thanks for testing those scenarios.

ssoloff commented:
So just to confirm what the current PBEM behavior is...

* Given
    * Player A emails player B a save game.
    * And this save game includes:
        * the PROVIDER (e.g. Gmail), SUBJECT, TO, and LOGIN configured by player A
        * a dummy PASSWORD

#### Scenario 1

1. When
    1. Player B has previously configured the same PROVIDER as player A (most likely with different values) and this configuration has been saved to player B's local cache.
    1. Player B starts a PBEM game.
    1. Player B opens the save game received from player A.
1. Then
    1. The PROVIDER from the save game is selected.
    1. SUBJECT and TO are populated from the save game.
    1. LOGIN and PASSWORD are populated from the local cache.

#### Scenario 2

1. When
    1. Player B has previously configured a different PROVIDER from player A and this configuration has been saved to player B's local cache.
    1. Player B starts a PBEM game.
    1. Player B opens the save game received from player A.
1. Then
    1. The PROVIDER from the save game is selected.
    1. SUBJECT, TO, and LOGIN are populated from the save game.
    1. PASSWORD is empty.

#### Scenario 3

1. When
    1. Player B has no PROVIDER configured.
    1. Player B starts a PBEM game.
    1. Player B opens the save game received from player A.
1. Then
    1. The PROVIDER from the save game is selected.
    1. SUBJECT, TO, and LOGIN are populated from the save game.
    1. PASSWORD is empty.

<hr>

Some observations...

The fact that PROVIDER and TO are always populated from the save game makes no sense to me.  PROVIDER should be whatever the local user has configured, and TO should really be populated from the sender's email, not who the sender sent the save game to (i.e. the receiver).  Whatever, though...  That's a usability issue and is outside the scope of this PR.  (I'm sure the same thing affects PBF, but there it doesn't matter because both players are, by definition, using the same forum provider and topic.)

LOGIN should **never** be populated from the save game.

Both LOGIN and PASSWORD should be populated from the local cache, if available.  Otherwise, empty values should be used.  (Only scenario A does not violate this requirement.)

<hr>

I'm going to ignore what I think _should_ be done in this PR, and simply try to mimic the current behavior with one exception...  As LOGIN is now encrypted, it will be impossible to ever use it to meaningfully populate the receiver's configuration in the event the receiver has no configuration in their local cache.  So I propose the following behavior for all scenarios:

* If the receiver has credentials for the PROVIDER in the save game, use those.
* If the receiver does not have credentials for the PROVIDER in the save game, use empty strings.
* No decryption errors (silent or otherwise) should be present in the console when loading a PBEM/PBF save game.

To achieve that last requirement, I'm probably going to have to avoid encrypting LOGIN and PASSWORD before they are written to the save game.  I'm hoping this will be straightforward as there is now a `credentialsProtected` flag I can use to control this behavior.

I'm also going to have to implement the privacy feature @panther2 and I discussed [above](https://github.com/triplea-game/triplea/pull/1955#issuecomment-309852082): use a dummy value for LOGIN in the save game just as a dummy value is used for PASSWORD.

ssoloff commented:
@ron-murhammer I believe everything is now working as expected with the latest commit (d2dbf62).

Here is a summary of the changes made in that commit:

* Credentials are no longer encrypted in save games--we simply store a dummy value for both user name and password, similar to what was done for the password alone prior to this PR.  Since the credentials are not encrypted, the receiver of the save game will not attempt to decrypt them, which ultimately failed because the sender and receiver have different encryption keys.
* To reiterate from the previous item: user name is no longer stored in the save game.  Based on @panther2's use case above, including it with the save game could lead to a privacy disclosure because some forums (e.g. AxisAndAllies.org) allow users to have a display name different from their login name.  A user may not appreciate us disclosing their login name for various reasons.
* The "Remember my password" option has been renamed "Remember me".  If not selected, both the user name and password will not be stored in the local cache.

I went through a battery of tests to ensure no errors were reported in the console and users never had to re-enter their credentials due to a compatibility issue.  The tests include:

* Save PBEM game using 3635 and loading it into this branch.
* Repeat previous test with a PBF game.
* Save PBEM game using this branch with one master password and loading it with a different master password (to simulate the save game being sent to a different user).
* Repeat previous test with a PBF game.
* Verify in a PBEM (Gmail) game that, even though the user name and password are removed from the save game, the user can still send the save game via email (i.e. the transient credentials are still present).
* Repeat previous test with a PBF (forums.triplea-game.org) game.

Note that I kind of hacked the multiple user tests.  In order to simulate multiple users, I created two different `local.cache` files and two different master passwords.  I would then swap in the appropriate `local.cache` and master password for each user depending on who I was "impersonating."  This is how I originally observed the tests fail earlier today, so I know it is reliable in that respect.

However, it would be nice to try a real end-to-end test.  If someone wants to volunteer, we can quickly try two or three turns of PBEM or PBF to ensure there are no issues when truly using different users and machines.

ssoloff commented:
If anyone is interested, I posted a PBF game [here](https://forums.triplea-game.org/post/1667).  Using the code from this branch, please play the German turn and re-post.  If you notice anything out of the ordinary, please comment here.

panther2 commented:
@ssoloff 

> However, it would be nice to try a real end-to-end test. If someone wants to volunteer, we can quickly try two or three turns of PBEM or PBF to ensure there are no issues when truly using different users and machines.

> If anyone is interested, I posted a PBF game here. Using the code from this branch, please play the German turn and re-post. If you notice anything out of the ordinary, please comment here.

I would be happy to assist testing, but would need a compiled version for doing so (I assume an updated triplea.jar would do it?).

ssoloff commented:
Thanks for offering to test, @panther2.  I uploaded binaries [here](https://github.com/ssoloff/triplea-game-triplea/releases/tag/1.9.0.0.private-1884).  Please let me know if you have any problems.

> I assume an updated triplea.jar would do it?

I believe that will work, as there are no installer changes in this PR.  But I uploaded installers, as well, just in case. :smile:

panther2 commented:
@ssoloff 

Thank you for providing the binaries, I have downloaded your above mentioned savegame. 
I reposted two times. The first repost has been done not activating "Remember Me" in the settings, the second repost included "Remember Me". In both cases I was not able to identify the unencrypted username and unencrypted password. Neither from the `local.cache` nor from any resulting `savegame.tsvg` .

So this appears to be the expected behaviour, unless I have not missed anything.

What I noticed was that after leaving the game, the PBEM settings window resetted to blank input fields, regardless whether I activated "Remember Me" or not. Only when loading the savegames the information "returned".

Do you want me to test anything else?

ssoloff commented:
Thanks, @panther2!

> What I noticed was that after leaving the game, the PBEM settings window resetted to blank input fields, regardless whether I activated "Remember Me" or not. Only when loading the savegames the information "returned".

Can you please list the exact steps you are taking here?  I did not notice that behavior during my testing.  I may have just overlooked it, or (more likely) you are doing something different in your testing than I did in mine that is triggering this behavior.

In the meantime, I will download your save game and make sure I don't have any problems on my end. :smile:

panther2 commented:
@ssoloff 

Please see:

![aaa](https://user-images.githubusercontent.com/20812775/27512147-3fe15592-5937-11e7-895f-9415fb4a92a0.gif)

ssoloff commented:
@panther2 Yes, I believe that behavior (resetting all fields in the PBEM/PBF panel to their default values) is by design (or if it's a bug, I believe it existed prior to this PR).  You'll notice that **File name** in the left-hand pane gets reset to the map location, not the save game location.  So it's no different than if you started a new PBEM/PBF game from scratch.

However, please confirm that after you return to the main menu from leaving the game, that if you change **Post to forum** from **disabled** to **forums.triplea-game.org**, the values are populated correctly.

BTW, I had no issues using either save game you posted.  So I would say we've successfully completed a round-trip test.  Thank you!

(Some bad news: If you now switch back to an earlier version, you will have to re-enter your credentials.  Sorry about that, I should have warned you earlier to backup your `local.cache` file.)

ssoloff commented:
@panther2 I confirmed the behavior you noted in your screencast also occurs in 1.9.0.0.3635, so it's not a regression.  As I said above, I think the code always resets back to "no save game loaded" when you return to the main menu, regardless of type of play.  Maybe the user experience could be better here; please feel free to open a new issue if you believe so.

ssoloff commented:
@ron-murhammer I think we are now ok functionally.  Please let me know if there's any other test scenarios I missed, or if you or @DanVanAtta have any feedback on the code.

ron-murhammer commented:
@ssoloff Looks good. And yeah the resetting back to the last selected map instead of save game is existing behavior which we could consider eventually changing. Nice work on this one.

panther2 commented:
@ssoloff 

> However, please confirm that after you return to the main menu from leaving the game, that if you change Post to forum from disabled to forums.triplea-game.org, the values are populated correctly.

Hereby confirmed!

> Some bad news: If you now switch back to an earlier version, you will have to re-enter your credentials. Sorry about that, I should have warned you earlier to backup your local.cache file.

No problem, I had anticipated that and backed it up before. Nevertheless I am always going with the latest prereleases, so it would have been a very temporary issue anyway...

> As I said above, I think the code always resets back to "no save game loaded" when you return to the main menu, regardless of type of play. Maybe the user experience could be better here; please feel free to open a new issue if you believe so.

I am not sure why I noticed it here and not before. I will go on with that current implementation monitoring my user experience ...

Thank you for all your work :smile:
